scripts/rollup/build.js

```diff
@@ -344,22 +385,30 @@ function getPlugins(
   const forks = Modules.getForks(bundleType, entry, moduleType, bundle);
   const isProduction = isProductionBundleType(bundleType);
   const isProfiling = isProfilingBundleType(bundleType);
-  const isUMDBundle =
-    bundleType === UMD_DEV ||
-    bundleType === UMD_PROD ||
-    bundleType === UMD_PROFILING;
-  const isFBWWWBundle =
-    bundleType === FB_WWW_DEV ||
-    bundleType === FB_WWW_PROD ||
-    bundleType === FB_WWW_PROFILING;
-  const isRNBundle =
-    bundleType === RN_OSS_DEV ||
-    bundleType === RN_OSS_PROD ||
-    bundleType === RN_OSS_PROFILING ||
-    bundleType === RN_FB_DEV ||
-    bundleType === RN_FB_PROD ||
-    bundleType === RN_FB_PROFILING;
-  const shouldStayReadable = isFBWWWBundle || isRNBundle || forcePrettyOutput;
+
+  const {isUMDBundle, shouldStayReadable} = getBundleTypeFlags(bundleType);
+
+  const needsMinifiedByClosure = isProduction && bundleType !== ESM_PROD;
+
+  // Any other packages that should specifically _not_ have sourcemaps
+  const sourcemapPackageExcludes = [
+    // Having `//#sourceMappingUrl` for the `react-debug-tools` prod bundle breaks
+    // `ReactDevToolsHooksIntegration-test.js`, because it changes Node's generated
+    // stack traces and thus alters the hook name parsing behavior.
+    // Also, this is an internal-only package that doesn't need sourcemaps anyway
+    'react-debug-tools',
+  ];
+
+  // Only generate sourcemaps for true "production" build artifacts
+  // that will be used by bundlers, such as `react-dom.production.min.js`.
+  // UMD and "profiling" builds are rarely used and not worth having sourcemaps.
+  const needsSourcemaps =
+    needsMinifiedByClosure &&
+    !isProfiling &&
+    !isUMDBundle &&
+    !sourcemapPackageExcludes.includes(entry) &&
+    !shouldStayReadable;
```

Comment:
why should a `shouldStayReadable` build not need sourcemaps? Are these not minified and name-mangled?

Reply:
Been a while since I've looked at the outputs, but my understanding is those _aren't_ minified, hence the "stay readable" part, and if it's not minified the existing bundle is acceptable without a sourcemap (especially since this PR implements sourcemaps _for_ the bundles right before they get minified anyway).

Reply:
ah! i was curious about what that meant *exactly*, was worried it was just prettified minified code 🙈